assets/hdfs_append/index.js

```diff
@@ -0,0 +1,161 @@
+'use strict';
+
+const _ = require('lodash');
+const Promise = require('bluebird');
+const path = require('path');
+
+
+function newProcessor(context, opConfig, jobConfig) {
+    // Client connection cannot be cached, an endpoint needs to be re-instantiated for a different
+    // namenode_host
+    opConfig.connection_cache = false;
+
+    /* This is used to keep track of HDFS append errors caused by corrupted replicas. If an error
+    /* occurs, a suffix will be added to the filename and incremented for each subsequent error. The
+    /* slice retry in a normal production job should take long enough to prevent the TS worker from
+    /* smashing the Namenode, but a timer mechanism will still need to be implemented in order to
+    /* prevent this from happening with faster-moving jobs. During the slice retry, this processor
+    /* will end up seeing the incremented filename in the `appendErrors` object and use the updated
+    /* name instead
+     */
+    // TODO: implement the timing mechanism. Current assumption is that retry and new slice
+    //       processing duration will keep the worker from toppling the Namenode
+    const appendErrors = {};
+
+    const clientService = getClient(context, opConfig, 'hdfs_ha');
+    const hdfsClient = clientService.client;
+
+    function prepareFile(filename, chunks) {
+        // We need to make sure the file exists before we try to append to it.
+        return hdfsClient.getFileStatusAsync(filename)
+            .catch(() => hdfsClient.mkdirsAsync(path.dirname(filename))
+                .then(() => hdfsClient.createAsync(filename, ''))
+                .catch((err) => {
+                    const errMsg = err.stack;
+                    return Promise.reject(
+                        `Error while attempting to create the file: ${filename} on hdfs, error: `
+                        + `${errMsg}`
+                    );
+                }))
+            .return(chunks)
+            // We need to serialize the storage of chunks so we run with concurrency 1
+            .map(chunks, chunk => hdfsClient.appendAsync(filename, chunk), { concurrency: 1 })
+            .catch((err) => {
+                const errMsg = err.stack ? err.stack : err;
+                let sliceError = '';
+                // Detecting the hdfs append error and updating the filename
+                if (errMsg.indexOf('AlreadyBeingCreatedException') > -1) {
+                    let newFilename = '';
+                    if (!appendErrors[filename]) {
```

Comment:
You might break this filename setup logic into a function so it's not complicating the error handling logic.